core/trino-main/src/main/java/io/trino/cost/JoinStatsRule.java

```diff
@@ -180,74 +182,61 @@ private PlanNodeStatsEstimate filterByEquiJoinClauses(
+                stats,
+                knownEstimates.stream().map(PlanNodeStatsEstimateWithClause::getEstimate).collect(toImmutableList()),
+                getJoinMultiClauseIndependenceFactor(session));
+        if (isNaN(outputRowCount)) {
+            return PlanNodeStatsEstimate.unknown();
         }
-        return filteredStats;
+        return normalizer.normalize(new PlanNodeStatsEstimate(outputRowCount, intersectCorrelatedJoinClause(stats, knownEstimates)), types);
     }
 
-    private PlanNodeStatsEstimate filterByAuxiliaryClause(PlanNodeStatsEstimate stats, EquiJoinClause clause, TypeProvider types)
+    private static Map<Symbol, SymbolStatsEstimate> intersectCorrelatedJoinClause(PlanNodeStatsEstimate stats, List<PlanNodeStatsEstimateWithClause> estimates)
     {
-        // we just clear null fraction and adjust ranges here
-        // selectivity is mostly handled by driving clause. We just scale heuristically by UNKNOWN_FILTER_COEFFICIENT here.
-
-        SymbolStatsEstimate leftStats = stats.getSymbolStatistics(clause.getLeft());
-        SymbolStatsEstimate rightStats = stats.getSymbolStatistics(clause.getRight());
-        StatisticRange leftRange = StatisticRange.from(leftStats);
-        StatisticRange rightRange = StatisticRange.from(rightStats);
-
-        StatisticRange intersect = leftRange.intersect(rightRange);
-        double leftFilterValue = firstNonNaN(leftRange.overlapPercentWith(intersect), 1);
-        double rightFilterValue = firstNonNaN(rightRange.overlapPercentWith(intersect), 1);
-        double leftNdvInRange = leftFilterValue * leftRange.getDistinctValuesCount();
-        double rightNdvInRange = rightFilterValue * rightRange.getDistinctValuesCount();
-        double retainedNdv = MoreMath.min(leftNdvInRange, rightNdvInRange);
-
-        SymbolStatsEstimate newLeftStats = buildFrom(leftStats)
-                .setNullsFraction(0)
-                .setStatisticsRange(intersect)
-                .setDistinctValuesCount(retainedNdv)
-                .build();
-
-        SymbolStatsEstimate newRightStats = buildFrom(rightStats)
-                .setNullsFraction(0)
-                .setStatisticsRange(intersect)
-                .setDistinctValuesCount(retainedNdv)
-                .build();
-
-        PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.buildFrom(stats)
-                .setOutputRowCount(stats.getOutputRowCount() * UNKNOWN_FILTER_COEFFICIENT)
-                .addSymbolStatistics(clause.getLeft(), newLeftStats)
-                .addSymbolStatistics(clause.getRight(), newRightStats);
-        return normalizer.normalize(result.build(), types);
+        PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder()
+                .addSymbolStatistics(stats.getSymbolStatistics());
+
+        for (PlanNodeStatsEstimateWithClause estimateWithClause : estimates) {
+            EquiJoinClause clause = estimateWithClause.getClause();
```

Comment:
did anything in logic changed compared to previous code?

Reply:
same logic inside the loop